## ANSWERS

#01. Importing Libraries

In [1]:
import pandas as pd
import numpy as np
import os

#02. Importing data set

In [2]:
path_name = r'/Users/geetika/Documents/Python/Instacart Basket Analysis/02_Data'

In [3]:
ords_prods_merge = pd.read_pickle(os.path.join(path_name, 'Prepared Data', 'orders_products_merged.pkl'))

In [4]:
ords_prods_merge.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2398795        1             2                   3                  7   
2    473747        1             3                   3                 12   
3   2254736        1             4                   4                  7   
4    431534        1             5                   4                 15   

   days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                     NaN         196                  1          0   
1                    15.0         196                  1          1   
2                    21.0         196                  1          1   
3                    29.0         196                  1          1   
4                    28.0         196                  1          1   

  product_name  aisle_id  department_id  prices _merge  
0         Soda        77              7     9.0   both  
1         Soda        77              7     9.0   both  
2         Soda        77              7     9.0   both  
3         Soda        77              7     9.0   both  
4         Soda        77              7     9.0   both

In [5]:
ords_prods_merge.shape

(32434212, 14)

In [6]:
# Find the aggregated mean of the "order_number" column gropued by "department_id" for the entire dataframe
ords_prods_merge.groupby('department_id').agg({'order_number':['mean']})

order_number
                      mean
department_id             
1                15.457687
2                17.277920
3                17.179756
4                17.811403
5                15.213779
6                16.439806
7                17.225773
8                15.340520
9                15.895474
10               20.197148
11               16.170828
12               15.887622
13               16.583304
14               16.757377
15               16.165037
16               17.663250
17               15.694469
18               19.310397
19               17.177343
20               16.473447
21               22.902379

In the exercise, we limited our analysis to the first one million entries, which is only a subset of the total dataset. A lot of the departments_ids were missing. Also not all order_numbers for the department_ids displayed are present. 

In the task, we are looking at the entire dataframe (32434212 entries). Therefore, the calculated means in the task differ from the exercise.

#03. Using transform() and loc() functions

In [7]:
# Using the transform() function to create the "max_order" column
ords_prods_merge['max_order'] = ords_prods_merge.groupby(['user_id'])['order_number'].transform(np.max)

In [8]:
ords_prods_merge.head(20)

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0    2539329        1             1                   2                  8   
1    2398795        1             2                   3                  7   
2     473747        1             3                   3                 12   
3    2254736        1             4                   4                  7   
4     431534        1             5                   4                 15   
5    3367565        1             6                   2                  7   
6     550135        1             7                   1                  9   
7    3108588        1             8                   1                 14   
8    2295261        1             9                   1                 16   
9    2550362        1            10                   4                  8   
10   2968173       15            15                   1                  9   
11   1870022       15            17                   2                 16   
12   1911383       15            18                   2                 11   
13   2715276       15            21                   1                  9   
14    487368       15            22                   1                 10   
15   2293453       19             2                   5                 14   
16   1973799       19             5                   6                 12   
17    532817       19             7                   4                 17   
18   1573906       21            10                   3                 10   
19   1593000       31            10                   3                  8   

    days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                      NaN         196                  1          0   
1                     15.0         196                  1          1   
2                     21.0         196                  1          1   
3                     29.0         196                  1          1   
4                     28.0         196                  1          1   
5                     19.0         196                  1          1   
6                     20.0         196                  1          1   
7                     14.0         196                  2          1   
8                      0.0         196                  4          1   
9                     30.0         196                  1          1   
10                     7.0         196                  2          0   
11                     8.0         196                  6          1   
12                     7.0         196                  1          1   
13                     7.0         196                  1          1   
14                    14.0         196                  1          1   
15                     6.0         196                  3          0   
16                     8.0         196                 15          1   
17                     6.0         196                  1          1   
18                     6.0         196                  2          0   
19                     7.0         196                 17          0   

   product_name  aisle_id  department_id  prices _merge  max_order  
0          Soda        77              7     9.0   both         10  
1          Soda        77              7     9.0   both         10  
2          Soda        77              7     9.0   both         10  
3          Soda        77              7     9.0   both         10  
4          Soda        77              7     9.0   both         10  
5          Soda        77              7     9.0   both         10  
6          Soda        77              7     9.0   both         10  
7          Soda        77              7     9.0   both         10  
8          Soda        77              7     9.0   both         10  
9          Soda        77              7     9.0   both         10  
10         Soda        77              7     9.0   both         22  
11         Soda        7

In [9]:
# Using the loc() functions to create the customer loyalty flag
ords_prods_merge.loc[ords_prods_merge['max_order'] > 40, 'loyalty_flag'] = 'Loyal customer'

In [10]:
ords_prods_merge.loc[(ords_prods_merge['max_order'] <= 40) & (ords_prods_merge['max_order'] > 10), 'loyalty_flag'] = 'Regular customer'

In [11]:
ords_prods_merge.loc[ords_prods_merge['max_order'] <= 10, 'loyalty_flag'] = 'New customer'

In [12]:
ords_prods_merge['loyalty_flag'].value_counts(dropna = False)

Regular customer    15891077
Loyal customer      10293737
New customer         6249398
Name: loyalty_flag, dtype: int64

In [13]:
ords_prods_merge.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2398795        1             2                   3                  7   
2    473747        1             3                   3                 12   
3   2254736        1             4                   4                  7   
4    431534        1             5                   4                 15   

   days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                     NaN         196                  1          0   
1                    15.0         196                  1          1   
2                    21.0         196                  1          1   
3                    29.0         196                  1          1   
4                    28.0         196                  1          1   

  product_name  aisle_id  department_id  prices _merge  max_order  \
0         Soda        77              7     9.0   both         10   
1         Soda        77              7     9.0   both         10   
2         Soda        77              7     9.0   both         10   
3         Soda        77              7     9.0   both         10   
4         Soda        77              7     9.0   both         10   

   loyalty_flag  
0  New customer  
1  New customer  
2  New customer  
3  New customer  
4  New customer

#04. Spending habits of the three types of customers

In [15]:
ords_prods_merge.groupby('loyalty_flag').agg({'prices':['mean', 'min', 'max']})

prices              
                       mean  min      max
loyalty_flag                             
Loyal customer    10.388747  1.0  99999.0
New customer      13.294370  1.0  99999.0
Regular customer  12.496203  1.0  99999.0

#05. Finding the different types of spenders

In [16]:
# Using the transform() function to create the "mean_price_of_products" column
ords_prods_merge['mean_price_of_products'] = ords_prods_merge.groupby(['user_id'])['prices'].transform(np.mean)

In [17]:
pd.options.display.max_rows = None

In [18]:
# Checking the aggregation results
ords_prods_merge.head(20)

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0    2539329        1             1                   2                  8   
1    2398795        1             2                   3                  7   
2     473747        1             3                   3                 12   
3    2254736        1             4                   4                  7   
4     431534        1             5                   4                 15   
5    3367565        1             6                   2                  7   
6     550135        1             7                   1                  9   
7    3108588        1             8                   1                 14   
8    2295261        1             9                   1                 16   
9    2550362        1            10                   4                  8   
10   2968173       15            15                   1                  9   
11   1870022       15            17                   2                 16   
12   1911383       15            18                   2                 11   
13   2715276       15            21                   1                  9   
14    487368       15            22                   1                 10   
15   2293453       19             2                   5                 14   
16   1973799       19             5                   6                 12   
17    532817       19             7                   4                 17   
18   1573906       21            10                   3                 10   
19   1593000       31            10                   3                  8   

    days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                      NaN         196                  1          0   
1                     15.0         196                  1          1   
2                     21.0         196                  1          1   
3                     29.0         196                  1          1   
4                     28.0         196                  1          1   
5                     19.0         196                  1          1   
6                     20.0         196                  1          1   
7                     14.0         196                  2          1   
8                      0.0         196                  4          1   
9                     30.0         196                  1          1   
10                     7.0         196                  2          0   
11                     8.0         196                  6          1   
12                     7.0         196                  1          1   
13                     7.0         196                  1          1   
14                    14.0         196                  1          1   
15                     6.0         196                  3          0   
16                     8.0         196                 15          1   
17                     6.0         196                  1          1   
18                     6.0         196                  2          0   
19                     7.0         196                 17          0   

   product_name  aisle_id  department_id  prices _merge  max_order  \
0          Soda        77              7     9.0   both         10   
1          Soda        77              7     9.0   both         10   
2          Soda        77              7     9.0   both         10   
3          Soda        77              7     9.0   both         10   
4          Soda        77              7     9.0   both         10   
5          Soda        77              7     9.0   both         10   
6          Soda        77              7     9.0   both         10   
7          Soda        77              7     9.0   both         10   
8          Soda        77              7     9.0   both         10   
9          Soda        77              7     9.0   both         10   
10         Soda        77              7     9.0   both         22   
11         S

In [19]:
# Checking the aggregation results for one user
ords_prods_merge[ords_prods_merge['user_id'] == 19]

order_id  user_id  order_number  orders_day_of_week  \
15         2293453       19             2                   5   
16         1973799       19             5                   6   
17          532817       19             7                   4   
2269387    2474845       19             6                   5   
3028920      86918       19             9                   5   
5101635    2293453       19             2                   5   
5101636    1973799       19             5                   6   
8116470     532817       19             7                   4   
8479132    2293453       19             2                   5   
9210215    2483716       19             8                   4   
9314597    2293453       19             2                   5   
9315663     532817       19             7                   4   
9430793      86918       19             9                   5   
10058025   2808909       19             4                   5   
10078217    532817       19             7                   4   
10199285   1321974       19             1                   6   
10199286   2293453       19             2                   5   
10200896   1321974       19             1                   6   
10200897   2293453       19             2                   5   
10200898   2808909       19             4                   5   
10200899   1973799       19             5                   6   
10201096   1321974       19             1                   6   
10201097   2808909       19             4                   5   
10201892   1321974       19             1                   6   
10201980   1321974       19             1                   6   
10202033   1321974       19             1                   6   
10202151   1321974       19             1                   6   
10202152   2474845       19             6                   5   
10202168   1321974       19             1                   6   
10202169   2474845       19             6                   5   
10202173   1321974       19             1                   6   
10202246   1321974       19             1                   6   
10202270   1321974       19             1                   6   
10202341   1321974       19             1                   6   
10202374   1321974       19             1                   6   
10202416   1321974       19             1                   6   
10202452   1321974       19             1                   6   
10202453   2293453       19             2                   5   
10202454   2208892       19             3                   0   
10202455   1973799       19             5                   6   
10202456   2474845       19             6                   5   
10202457    532817       19             7                   4   
10202458   2483716       19             8                   4   
10202459     86918       19             9                   5   
10209569   1321974       19             1                   6   
10209570   2293453       19             2                   5   
10209571   2808909       19             4                   5   
10211665   1321974       19             1                   6   
10211666   2293453       19             2                   5   
10211667   2808909       19             4                   5   
10212714   1321974       19             1                   6   
10212715   2293453       19             2                   5   
10212716   2208892       19             3                   0   
10212717   1973799       19             5                   6   
10213129   1321974       19             1                   6   
10213621   1321974       19             1                   6   
10219416   1321974       19             1                   6   
10219417   2293453       19             2                   5   
10219418   2208892       19             3                   0   
10219419   1973799       19             5                   6   
10219420   2474845       19             6    

In [20]:
ords_prods_merge.loc[ords_prods_merge['mean_price_of_products'] < 10, 'spending_flag'] = 'Low spender'

In [21]:
ords_prods_merge.loc[ords_prods_merge['mean_price_of_products'] >= 10, 'spending_flag'] = 'High spender'

In [22]:
ords_prods_merge['spending_flag'].value_counts(dropna = False)

Low spender     31798751
High spender      635461
Name: spending_flag, dtype: int64

#05. Finding the regularity of a user's ordering behaviour

In [23]:
# Using the transform() function to create the "mean_days_since_prior_order" column
ords_prods_merge['median_days_since_prior_order'] = ords_prods_merge.groupby(['user_id'])['days_since_prior_order'].transform(np.median)

In [24]:
ords_prods_merge.head(100)

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0    2539329        1             1                   2                  8   
1    2398795        1             2                   3                  7   
2     473747        1             3                   3                 12   
3    2254736        1             4                   4                  7   
4     431534        1             5                   4                 15   
5    3367565        1             6                   2                  7   
6     550135        1             7                   1                  9   
7    3108588        1             8                   1                 14   
8    2295261        1             9                   1                 16   
9    2550362        1            10                   4                  8   
10   2968173       15            15                   1                  9   
11   1870022       15            17                   2                 16   
12   1911383       15            18                   2                 11   
13   2715276       15            21                   1                  9   
14    487368       15            22                   1                 10   
15   2293453       19             2                   5                 14   
16   1973799       19             5                   6                 12   
17    532817       19             7                   4                 17   
18   1573906       21            10                   3                 10   
19   1593000       31            10                   3                  8   
20   2231262       31            17                   3                 11   
21   2580647       43             6                   4                 16   
22   2187180       43             9                   4                 12   
23   2497897       52             1                   1                  9   
24   1318871       52             2                   1                 10   
25   1261384       52             3                   1                 10   
26    580568       52             5                   2                 10   
27   2428073       52             7                   2                  7   
28    342306       52             9                   1                 11   
29   1520353       52            10                   2                 10   
30   2542086       52            11                   2                  9   
31    944694       52            12                   1                 16   
32   1268191       52            15                   2                 11   
33   1122089       52            17                   2                 10   
34   1498922       52            18                   2                 10   
35    180919       52            19                   3                  8   
36   2695875       52            21                   5                 10   
37   3205960       67             1                   3                 11   
38    970922       67             2                   1                 15   
39    596585       67             3                   1                 13   
40   2037087       67             4                   2                  9   
41   2474111       67             5                   3                 14   
42   1568481       67             6                   3                 12   
43   2048527       67             8                   2                 11   
44   2389031       67             9                   2                 14   
45   2378562       67            10                   1                 10   
46   1515698       67            11                   1                 11   
47   1928050       67            12                   2                 12   
48   1949274       67            13                   1                 10   
49    551869       67            14                   1                  9   
50    768572       67     

In [25]:
ords_prods_merge['median_days_since_prior_order'].value_counts(dropna = False)

7.0     5689347
6.0     2985491
8.0     2714204
5.0     2490725
4.0     2315881
30.0    1684596
9.0     1638069
3.0     1482100
10.0    1273391
14.0    1214744
13.0    1058996
11.0    1051348
12.0     973920
15.0     744545
2.0      561620
16.0     515980
21.0     422311
17.0     399621
20.0     371620
18.0     357965
19.0     349646
22.0     265921
23.0     202410
28.0     200284
27.0     173635
24.0     169245
26.0     161998
25.0     152541
1.0      151844
29.0     137138
7.5       49560
6.5       43406
8.5       37705
0.0       34237
9.5       31239
10.5      30377
5.5       29888
4.5       28269
12.5      25556
13.5      23107
11.5      20079
14.5      18494
15.5      14682
18.5      14467
17.5      12823
3.5       11762
20.5       9893
16.5       9722
19.5       9442
25.5       7834
22.5       7680
2.5        7354
24.5       7323
21.5       7093
28.5       6781
23.5       6777
27.5       5710
26.5       5482
29.5       5017
1.5        1058
0.5         259
Name: median_days_since_

In [26]:
# Using the loc() function to create the "Regularity_of_customer" column
ords_prods_merge.loc[ords_prods_merge['median_days_since_prior_order'] > 20, 'Regularity_of_customer'] = 'Non-frequent customer'

In [27]:
ords_prods_merge.loc[(ords_prods_merge['median_days_since_prior_order'] <= 20) & (ords_prods_merge['median_days_since_prior_order'] > 10), 'Regularity_of_customer'] = 'Regular customer'

In [28]:
ords_prods_merge.loc[ords_prods_merge['median_days_since_prior_order'] <= 10, 'Regularity_of_customer'] = 'Frequent customer'

In [29]:
ords_prods_merge['Regularity_of_customer'].value_counts(dropna = False)

Frequent customer        21577409
Regular customer          7217134
Non-frequent customer     3639669
Name: Regularity_of_customer, dtype: int64

In [31]:
# Exporting the dataframe
ords_prods_merge.to_pickle(os.path.join(path_name, 'Prepared Data', 'ords_prods_merge_E4.8.pkl'))